### DISCUSSION TITLE
Adjacency list relationships with aliased join

### DISCUSSION THREAD

**Discussion by minyoung90** [sqlalchemy/sqlalchemy#8026] [Usage Questions] [answered]
Hello, I have a problem with below structure...

```python
# Model
class Node(Base):
    __tablename__ = 'node'
    id = Column(Integer, primary_key=True)
    parent_id = Column(Integer, ForeignKey('node.id'))
    parent = relationship(
        "Node", back_populates="children"
    )
    data = Column(String(50))  
    children = relationship("Node")

# aliasing...
fst_depth = aliased(Node)
snd_depth = aliased(Node)
trd_depth = aliased(Node)

# query
stmt = select(fst_depth)
           .outerjoin(snd_depth, fst_depth.id == snd_depth.parent_id)
           .outerjoin(trd_depth, snd_depth.id == trd_depth.parent_id)
           .options(
                contains_eager(
                    fst_depth.children.of_type(snd_depth)
                )
           )
print(stmt)
```

What I expected was that "snd_depth" should be chained to "fst_depth" and "trd_depth" chained to "snd_depth" vise versa.
But, above stmt gave me that "snd_depth" was chained to "fst_depth" but "trd_depth" chained to "fst_depth" too.
In other words, this condition of "fst_depth.id == trd_depth.parent_id" is in sqlalchemy query log.

What am I wrong?

**Comment by zzzeek** [comment, marked as answer]
hi - 

I ran your script and it outputs:

```sql
SELECT node_1.id, node_1.parent_id, node_1.data 

      FROM node AS node_1 
      LEFT OUTER JOIN node AS node_2 ON node_1.id = node_2.parent_id 
      LEFT OUTER JOIN node AS node_3 ON node_2.id = node_3.parent_id

```

looks fine to me, is that what you see?

**Reply by minyoung90** [reply]
Oh... sorry. I updated my question. I added "options" to query before. that options (expecially, contains_eager) changes query.

**Reply by zzzeek** [reply]
OK, the query is now:

```sql
    SELECT node_1.id, node_1.parent_id, node_1.data, node_2.id AS id_1, node_2.parent_id AS parent_id_1, node_2.data AS data_1 
    FROM node AS node_2 
    LEFT OUTER JOIN node AS node_1 ON node_2.id = node_1.parent_id 
    LEFT OUTER JOIN node AS node_3 ON node_1.id = node_3.parent_id
```

so what contains_eager did is that it added columns from both the first alias, and the second, to the columns clause of the SELECT.   As it happens, the current implementation rendered the second alias's columns first, so the anonymous naming heuristic that produces names like "node_1" and "node_2" reversed the names used from the previous query.  Whereas the previous query SELECTed from "node_1 JOIN node_2 JOIN node_3", this one is doing "node_2 JOIN node_1 JOIN node_3", which might be making it harder to read.  To make it easier to read, give your aliases names:

```py
fst_depth = aliased(Node, name="fst")
snd_depth = aliased(Node, name="snd")
trd_depth = aliased(Node, name="trd")
```

then it prints:

```sql
SELECT snd.id, snd.parent_id, snd.data, fst.id AS id_1, fst.parent_id AS parent_id_1, fst.data AS data_1 
FROM node AS fst 
LEFT OUTER JOIN node AS snd ON fst.id = snd.parent_id 
LEFT OUTER JOIN node AS trd ON snd.id = trd.parent_id
```

which is easier to read.

**Reply by minyoung90** [reply]
@zzzeek Thanks!! It makes me right to see the query... but ... I have another problem....

```python
fst_depth = aliased(Node, name="fst")
snd_depth = aliased(Node, name="snd")
trd_depth = aliased(Node, name="trd")

stmt = select(fst_depth)
           .outerjoin(snd_depth, fst_depth.id == snd_depth.parent_id)
           .outerjoin(trd_depth, snd_depth.id == trd_depth.parent_id)
           .options(
                contains_eager(
                    fst_depth.children.of_type(snd_depth)
                ),
                contains_eager(
                    snd_depth.children.of_type(trd_depth)
                ),
           )
```

If I added snd_depth of contains_eager then, I got this error message...

```
sqlalchemy.exc.ArgumentError: Mapped attribute "Node.children" does not apply to any of the root entities in this query, e.g. aliased(Node). Please specify the full path from one of the root entities to the target attribute.
```

Does not guarantee that aliased makes full path of target?

**Reply by minyoung90** [reply]
If I edit above like this...

```python
stmt = select(fst_depth)
           .outerjoin(snd_depth, fst_depth.id == snd_depth.parent_id)
           .outerjoin(trd_depth, snd_depth.id == trd_depth.parent_id)
           .options(
                contains_eager(
                    fst_depth.children.of_type(snd_depth)
                ),
                contains_eager(
                    fst_depth.children.of_type(snd_depth),
                    snd_depth.children.of_type(trd_depth)
                ),
           )
```

Every things fine!!

**Reply by zzzeek** [reply]
so the phrase "does not apply to any of the root entities " uses this term "root entity", which granted is not any kind of formal term.  

What it means is if you have a query like:

     SELECT * FROM a JOIN b ON ... JOIN c ON ...

the "root" entities above is "a".

For this query:

     SELECT * FROM a, b JOIN c ON ...

the "root" entities is "a", "b".   

it means the FROM clauses that follow the FROM keyword, that aren't linked by JOIN.    

so in the query you have, "snd_depth" is not a "root" entity, it's JOINed to "fst_depth".     contains_eager needs to be spelled out in such a way that mirrors the structure of the joins, which here is most succinctly:


```py

stmt = select(fst_depth)
           .outerjoin(snd_depth, fst_depth.id == snd_depth.parent_id)
           .outerjoin(trd_depth, snd_depth.id == trd_depth.parent_id)
           .options(
                contains_eager(
                    fst_depth.children.of_type(snd_depth)
                ).contains_eager(
                    snd_depth.children.of_type(trd_depth)
                ),
           )
```

where note the above is exactly what you had except we link the two contains_eager() with a dot.

the form `contains_eager(A.b).contains_eager(B.c)` as we have above is, as you noted in your other answer, the modern form for this:

```py
# longhand for `contains_eager(A.b).contains_eager(B.c)`

contains_eager(A.b),
contains_eager(A.b, B.c)
```

That syntax is still supported but not really documented anymore as it is not as simple as the method chained syntax, which is discussed at https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#relationship-loading-with-loader-options .
